### PR TITLE
Refactor workers

### PR DESCRIPTION
**Pull request by shimonewman**

### CODE REVIEW

File: app/controllers/api/functions.php [hunk 1]
```diff
@@ -158,10 +156,11 @@
     ->inject('dbForProject')
     ->inject('project')
     ->inject('user')
-    ->inject('events')
+    ->inject('queueForEvents')
+    ->inject('queueForBuilds')
```
Comment (christyjacob4) [inline app/controllers/api/functions.php:160]: I don't see these resources created in the diff of the `init.php` file. Can you check if these have been added?

File: app/controllers/api/account.php [hunk 2]
```diff
@@ -897,9 +898,9 @@
     ->inject('project')
     ->inject('dbForProject')
     ->inject('locale')
-    ->inject('events')
-    ->inject('mails')
-    ->action(function (string $userId, string $email, string $url, Request $request, Response $response, Document $user, Document $project, Database $dbForProject, Locale $locale, Event $events, Mail $mails) {
+    ->inject('queueForEvents')
+    ->inject('queueForMails')
```
Comment (christyjacob4) [inline app/controllers/api/account.php:902]: I don't see these resources being declared in the init.php file

File: composer.json [hunk 3]
```diff
@@ -56,13 +56,13 @@
...
+        "utopia-php/messaging": "0.2.*",
         "utopia-php/migration": "0.3.*",
         "utopia-php/orchestration": "0.9.*",
-        "utopia-php/platform": "0.4.*",
+        "utopia-php/platform": "dev-integrate-workers as 0.3.3",
         "utopia-php/pools": "0.4.*",
         "utopia-php/preloader": "0.2.*",
-        "utopia-php/queue": "0.5.*",
+        "utopia-php/queue": "dev-feat-get-worker-start as 0.5.3",
         "utopia-php/registry": "0.5.*",
```
Comment (christyjacob4) [inline composer.json:66]: We need to get the proper library versions released before merging to 1.4.x

File: docker-compose.yml [hunk 4]
```diff
@@ -83,7 +83,10 @@ services:
       - ./docs:/usr/src/code/docs
       - ./public:/usr/src/code/public
       - ./src:/usr/src/code/src
-      - ./dev:/usr/src/code/dev
+      - ./dev:/usr/local/dev
+      - ./vendor/utopia-php/platform:/usr/src/code/vendor/utopia-php/platform
+      - ./vendor/utopia-php/framework:/usr/src/code/vendor/utopia-php/framework
```
Comment (christyjacob4) [inline docker-compose.yml:88]: We've to remove all vendor mounts from the docker compose file

File: docker-compose.yml [hunk 5]
```diff
@@ -378,7 +390,8 @@ services:
       - _APP_DB_USER
       - _APP_DB_PASS
       - _APP_LOGGING_PROVIDER
-      - _APP_LOGGING_CONFIG
```
Comment (christyjacob4) [inline docker-compose.yml:381]: this variable is required and should not be removed

File: src/Appwrite/Event/Audit.php [hunk 6]
```diff
@@ -11,9 +12,14 @@ class Audit extends Event
...
 
-    public function __construct()
+    public function __construct(protected Connection $connection)
     {
-        parent::__construct(Event::AUDITS_QUEUE_NAME, Event::AUDITS_CLASS_NAME);
+        parent::__construct($connection);
+
+        $this
+            ->setQueue(Event::AUDITS_QUEUE_NAME)
+            ->setClass(Event::BUILDS_CLASS_NAME);
```
Comment (christyjacob4) [inline src/Appwrite/Event/Audit.php:21]:
```suggestion
            ->setClass(Event::AUDITS_CLASS_NAME);
```

File: .gitmodules [hunk 7]
```diff
@@ -1,4 +1,4 @@
 [submodule "app/console"]
 	path = app/console
 	url = https://github.com/appwrite/console
-	branch = 3.2.1
+	branch = 3.2.2
```
Comment (christyjacob4) [inline .gitmodules:4]: This change shouldn't be part of this PR

File: app/controllers/api/account.php (hunk 2 above)
Comment (shimonewman) [reply app/controllers/api/account.php:902]: they are

File: app/controllers/api/functions.php (hunk 1 above)
Comment (shimonewman) [reply app/controllers/api/functions.php:160]: they are

File: composer.json (hunk 3 above)
Comment (shimonewman) [reply composer.json:66]: Cool

File: src/Appwrite/Event/Database.php [hunk 8]
```diff
@@ -12,9 +14,11 @@ class Database extends Event
     protected ?Document $collection = null;
     protected ?Document $document = null;
 
-    public function __construct()
+    public function __construct(protected Connection $connection)
     {
-        parent::__construct(Event::DATABASE_QUEUE_NAME, Event::DATABASE_CLASS_NAME);
+        parent::__construct($connection);
+
+        $this->setClass(Event::DATABASE_CLASS_NAME);
```
Comment (christyjacob4) [inline src/Appwrite/Event/Database.php:21]: the queue does not need to be set here ?

File: src/Appwrite/Platform/Workers/Certificates.php [hunk 9]
```diff
@@ -445,30 +484,28 @@ private function updateDomainDocuments(string $certificateId, string $domain, bo
...
                 ->setParam('ruleId', $rule->getId())
                 ->setPayload($rule->getArrayCopy(array_keys($ruleModel->getRules())))
                 ->trigger();
 
+
             /** Trigger Functions */
-            $ruleUpdate
-                ->setClass(Event::FUNCTIONS_CLASS_NAME)
-                ->setQueue(Event::FUNCTIONS_QUEUE_NAME)
+            $queueForFunctions
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Certificates.php:500]: $queueForFunctions does not have project, event, param, payload set.

File: src/Appwrite/Platform/Workers/Databases.php [hunk 10]
```diff
@@ -1,76 +1,94 @@
...
-        $type = $this->args['type'];
-        $project = new Document($this->args['project']);
-        $collection = new Document($this->args['collection'] ?? []);
-        $document = new Document($this->args['document'] ?? []);
-        $database = new Document($this->args['database'] ?? []);
+        $this
+            ->desc('Databases worker')
+            ->inject('message')
+            ->inject('dbForConsole')
+            ->inject('dbForProject')
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Databases.php:34]: Are you sure you can directly use `dbForProject` here? 

What is the `projectId` when this resource is created? From what I remember in the code ,we dont know the project ID until we parse `$this->args['project']`

File: app/cli.php [hunk 11]
```diff
@@ -17,10 +19,13 @@
 use Utopia\Database\Document;
 use Utopia\Logger\Log;
 use Utopia\Pools\Group;
+use Utopia\Queue\Connection;
 use Utopia\Registry\Registry;
 
 Authorization::disable();
 
+global $register;
```
Comment (PineappleIOnic) [inline app/cli.php:27]: Why are we adding a global here? Shouldn't getResource('register') be all we need?

File: src/Appwrite/Platform/Workers/Messaging.php [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+        $from = App::getEnv('_APP_SMS_FROM');
+
+        if (empty($from)) {
+            Console::error('Skipped sms processing. No phone number has been set.');
+            return;
+        }
+
+        $message = new SMS(
+            to: [$payload['recipient']],
```
Comment (christyjacob4) [inline src/Appwrite/Platform/Workers/Messaging.php:96]: Are you sure $to is an array ?

File: src/Appwrite/Platform/Tasks/CalcTierStats.php [hunk 13]
```diff
@@ -155,12 +155,12 @@ public function action(Group $pools, Cache $cache, Database $dbForConsole, Regis
                     }
 
                     /** Get Usage stats */
-                    $range = '90d';
+                    $range = '30d';
```
Comment (PineappleIOnic) [inline src/Appwrite/Platform/Tasks/CalcTierStats.php:158]: Was there a reason this was reduced?

File: src/Appwrite/Platform/Workers/Databases.php [hunk 14]
```diff
@@ -1,76 +1,94 @@
 <?php
 
+namespace Appwrite\Platform\Workers;
+
 use Appwrite\Event\Event;
-use Appwrite\Extend\Exception;
 use Appwrite\Messaging\Adapter\Realtime;
-use Appwrite\Resque\Worker;
+use Exception;
```
Comment (PineappleIOnic) [inline src/Appwrite/Platform/Workers/Databases.php:7]: Shouldn't we be using \Exception when we need this?

**Review by PineappleIOnic — COMMENTED** [review]
Just a few things, but everything else looks good

File: src/Appwrite/Event/Migration.php [hunk 15]
```diff
@@ -89,10 +98,11 @@ public function trigger(): string|bool
...
-            'project' => $this->project,
-            'user' => $this->user,
-            'migration' => $this->migration
-        ]);
+        return;
+//        ResqueScheduler::enqueueAt($at, $this->queue, $this->class, [
+//            'project' => $this->project,
+//            'user' => $this->user,
+//            'migration' => $this->migration
+//        ]);
```
[the remark this answers is not in the document]
Comment (shimonewman) [reply src/Appwrite/Event/Migration.php:106]: This is a copy paste thing (not in use)

File: app/views/install/compose.phtml [hunk 16]
```diff
@@ -345,6 +345,8 @@ services:
       - _APP_DB_PASS
       - _APP_LOGGING_PROVIDER
       - _APP_LOGGING_CONFIG
+      - _APP_WORKERS_NUM=1
+      - _APP_QUEUE_NAME=database_db_main
```
Comment (christyjacob4) [inline app/views/install/compose.phtml:349]: please double check the name of the default database. Im not sure its `db_main`

File: app/worker.php [hunk 17]
```diff
@@ -72,61 +113,150 @@
...
+}, ['queue']);
+Server::setResource('queueForFunctions', function (Connection $queue) {
+    return new Func($queue);
+}, ['queue']);
+Server::setResource('queueForCertificates', function (Connection $queue) {
+    return new Certificate($queue);
+}, ['queue']);
+Server::setResource('usage', function ($register) {
+    return new Stats($register->get('statsd'));
+}, ['register']);
```
Comment (christyjacob4) [inline app/worker.php:153]: Can you move the `usage` and `logger` resources to the top so all the `queue*` resources are together at the bottom?

File: src/Appwrite/Platform/Workers/Databases.php (hunk 10 above)
Comment (shimonewman) [reply src/Appwrite/Platform/Workers/Databases.php:34]: dbForProject is with the project arg

Server::setResource('dbForProject', function (Cache $cache, Registry $register, Message $message, Database $dbForConsole) {
    $payload = $message->getPayload() ?? [];
    $project = new Document($payload['project'] ?? []);

File: src/Appwrite/Platform/Workers/Messaging.php [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+            Console::error('Skipped sms processing. No Phone provider has been set.');
+            return;
+        }
+
+        $from = App::getEnv('_APP_SMS_FROM');
+
+        if (empty($from)) {
+            Console::error('Skipped sms processing. No phone number has been set.');
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (shimonewman) [reply src/Appwrite/Platform/Workers/Messaging.php:93]: ok

File: app/console
Comment (christyjacob4) [inline app/console:1]: The submodule has been deleted. you need to revert his back

File: src/Appwrite/Platform/Workers/Messaging.php (hunk 12 above)
Comment (shimonewman) [reply src/Appwrite/Platform/Workers/Messaging.php:96]: I believe so 
[](https://github.com/appwrite/appwrite/blob/1.4.x/app/workers/messaging.php#L63)

Comment (shimonewman) [reply src/Appwrite/Platform/Workers/Messaging.php:96]: image

File: src/Appwrite/Platform/Tasks/CalcTierStats.php (hunk 13 above)
Comment (christyjacob4) [reply src/Appwrite/Platform/Tasks/CalcTierStats.php:158]: Yea this is for the hamster script. Its alright

File: app/worker.php (hunk 17 above)
Comment (shimonewman) [reply app/worker.php:153]: cool

File: app/views/install/compose.phtml (hunk 16 above)
Comment (shimonewman) [reply app/views/install/compose.phtml:349]: it is

File: src/Appwrite/Platform/Workers/Databases.php (hunk 14 above)
Comment (shimonewman) [reply src/Appwrite/Platform/Workers/Databases.php:7]: Why not import it instead?

File: app/worker.php [hunk 19]
```diff
@@ -160,3 +290,14 @@
...
     });
+
+try {
+    $workerStart = $worker->getWorkerStart();
+} catch (\Throwable $error) {
+     $worker->workerStart()
+         ->action(function () use ($workerName) {
+             Console::info("Worker $workerName  started");
+         });
+}
```
Comment (christyjacob4) [inline app/worker.php:301]: what exactly  is this code doing?

File: app/worker.php [hunk 20]
```diff
@@ -72,61 +113,150 @@
...
     });
 
-$server
+$worker
     ->error()
     ->inject('error')
     ->inject('logger')
     ->inject('log')
-    ->action(function (Throwable $error, ?Logger $logger, Log $log) {
+    ->action(function (Throwable $error, Logger|null $logger, Log $log) {
```
Comment (christyjacob4) [inline app/worker.php:259]:
```suggestion
    ->action(function (Throwable $error, ?Logger $logger, Log $log) {
```
What is the difference between the two?

File: app/cli.php (hunk 11 above)
Comment (shimonewman) [reply app/cli.php:27]: It does; it is mainly for the editor.
Removed it

File: src/Appwrite/Platform/Workers/Messaging.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (shimonewman) [reply src/Appwrite/Platform/Workers/Messaging.php:93]: fixed

File: src/Appwrite/Platform/Workers/Certificates.php (hunk 9 above)
Comment (shimonewman) [reply src/Appwrite/Platform/Workers/Certificates.php:500]: fixed

File: src/Appwrite/Platform/Workers/Messaging.php (hunk 12 above)
Comment (christyjacob4) [reply src/Appwrite/Platform/Workers/Messaging.php:96]: Sounds good 👍

File: app/views/install/compose.phtml (hunk 16 above)
Comment (christyjacob4) [reply app/views/install/compose.phtml:349]: lets remove these variables from here and hard code them as default values in the code

File: composer.json (hunk 3 above)
Comment (shimonewman) [reply composer.json:66]: fixed

File: docker-compose.yml (hunk 4 above)
Comment (shimonewman) [reply docker-compose.yml:88]: fixed

File: app/worker.php (hunk 17 above)
Comment (shimonewman) [reply app/worker.php:153]: fixed

File: app/cli.php [hunk 21]
```diff
@@ -140,10 +143,18 @@
...
+CLI::setResource('queue', function (Group $pools) {
+    return $pools->get('queue')->pop()->getResource();
+}, ['pools']);
 CLI::setResource('queueForFunctions', function (Group $pools) {
     return new Func($pools->get('queue')->pop()->getResource());
 }, ['pools']);
-
+CLI::setResource('queueForDeletes', function (Connection $queue) {
+    return new Delete($queue);
+}, ['queue']);
```
[the remark this answers is not in the document]
Comment (shimonewman) [reply app/cli.php:154]: fixed

File: app/worker.php [hunk 22]
```diff
@@ -72,56 +113,145 @@
...
 Server::setResource('log', fn() => new Log());
 
-Server::setResource('logger', function ($register) {
-    return $register->get('logger');
-}, ['register']);
+/**
+ * Get Functions Storage Device
+ * @param string $projectId of the project
+ * @return Device
+ */
```
[the remark this answers is not in the document]
Comment (shimonewman) [reply app/worker.php:171]: fixed

File: app/worker.php [hunk 23]
```diff
@@ -72,56 +113,145 @@
...
 }
 
-$adapter = new Swoole($connection, $workerNumber, App::getEnv('QUEUE'));
-$server = new Server($adapter);
+\array_shift($args);
+$workerName = $args[0];
+$workerIndex = $args[1] ?? '';
 
-$server
+if (!empty($workerNum)) {
```
[the remark this answers is not in the document]
Comment (shimonewman) [reply app/worker.php:224]: fixed

File: app/worker.php [hunk 24]
```diff
@@ -72,56 +113,145 @@
...
+if (!isset($args[1])) {
+    Console::error('Missing worker name');
+    Console::exit(1);
 }
 
-$adapter = new Swoole($connection, $workerNumber, App::getEnv('QUEUE'));
-$server = new Server($adapter);
+\array_shift($args);
+$workerName = $args[0];
+$workerIndex = $args[1] ?? '';
```
[the remark this answers is not in the document]
Comment (shimonewman) [reply app/worker.php:194]: via the command line when the container is created , like so: php /usr/src/code/app/worker.php functions $@

File: src/Appwrite/Event/Event.php [hunk 25]
```diff
@@ -3,12 +3,13 @@
...
 use InvalidArgumentException;
-use Resque;
 use Utopia\Database\Document;
+use Utopia\Queue\Client;
+use Utopia\Queue\Connection;
 
 class Event
 {
-    public const DATABASE_QUEUE_NAME = 'v1-database';
+    public const DATABASE_QUEUE_NAME = 'v1-databases';
```
[the remark this answers is not in the document]
Comment (shimonewman) [reply src/Appwrite/Event/Event.php]: fixed

File: docker-compose.yml [hunk 26]
```diff
@@ -379,6 +391,8 @@ services:
       - _APP_DB_PASS
       - _APP_LOGGING_PROVIDER
       - _APP_LOGGING_CONFIG
+      - _APP_WORKERS_NUM=1
+      - _APP_QUEUE_NAME=database_db_main
```
[the remark this answers is not in the document]
Comment (shimonewman) [reply docker-compose.yml:383]: fixed

File: app/views/install/compose.phtml (hunk 16 above)
[the remark this answers is not in the document]
Comment (shimonewman) [reply app/views/install/compose.phtml:349]: fixed

**Review by eldadfux — APPROVED** [review]

File: composer.json [hunk 27]
```diff
@@ -59,7 +59,7 @@
         "utopia-php/messaging": "0.1.*",
         "utopia-php/migration": "0.3.*",
         "utopia-php/orchestration": "0.9.*",
-        "utopia-php/platform": "0.4.*",
+        "utopia-php/platform": "0.5.*",
         "utopia-php/pools": "0.4.*",
         "utopia-php/preloader": "0.2.*",
         "utopia-php/queue": "0.5.*",
```
[the remark this answers is not in the document]
Comment (christyjacob4) [reply composer.json:65]: removed this in another PR
https://github.com/appwrite/appwrite/pull/6946

File: app/cli.php [hunk 28]
```diff
@@ -140,10 +141,18 @@
...
 
-CLI::setResource('queueForFunctions', function (Group $pools) {
-    return new Func($pools->get('queue')->pop()->getResource());
+CLI::setResource('queue', function (Group $pools) {
+    return $pools->get('queue')->pop()->getResource();
 }, ['pools']);
-
+CLI::setResource('queueForFunctions', function (Connection $queue) {
+    return new Func($queue);
+}, ['pools']);
```
Comment (stnguyen90) [inline app/cli.php:149]: should this be `queue`?

File: app/views/install/compose.phtml [hunk 29]
```diff
@@ -188,7 +188,6 @@ services:
       - traefik.http.routers.appwrite_realtime_wss.rule=PathPrefix(`/v1/realtime`)
       - traefik.http.routers.appwrite_realtime_wss.service=appwrite_realtime
       - traefik.http.routers.appwrite_realtime_wss.tls=true
-      - traefik.http.routers.appwrite_realtime_wss.tls.certresolver=dns
```
Comment (stnguyen90) [inline app/views/install/compose.phtml:191]: what's the reason for removing this?

File: app/worker.php (hunk 24 above)
[the remark this answers is not in the document]
Comment (stnguyen90) [reply app/worker.php:197]: Where exactly? Because I don't see anything specified anywhere like:
 https://github.com/appwrite/appwrite/blob/20985c7ee4764d8266decae44e58068cc4168263/app/views/install/compose.phtml#L214-L241

File: docker-compose.yml [hunk 30]
```diff
@@ -379,6 +378,8 @@ services:
       - _APP_DB_PASS
       - _APP_LOGGING_PROVIDER
       - _APP_LOGGING_CONFIG
+      - _APP_WORKERS_NUM
+      - _APP_QUEUE_NAME
```
Comment (stnguyen90) [inline docker-compose.yml:382]: Shouldn't these be added to app/config/variables.php? or some .env file?

**Review by christyjacob4 — APPROVED** [review]

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

File: app/views/install/compose.phtml (hunk 29 above)
Comment (christyjacob4) [reply app/views/install/compose.phtml:191]: This is not being used. We have not declared a resolver named `dns` in the traefik configuration above. 
And traefik complains about it too.

File: docker-compose.yml (hunk 30 above)
Comment (christyjacob4) [reply docker-compose.yml:382]: We don't want to introduce these on self-hosted yet since these are temporary.

File: app/cli.php (hunk 28 above)
Comment (christyjacob4) [reply app/cli.php:149]: Good catch 👍

File: docker-compose.yml (hunk 30 above)
Comment (christyjacob4) [reply docker-compose.yml:382]: Screenshot 2023-10-19 at 1 31 46 AM
Here, we pass the args ^

**Review by stnguyen90 — APPROVED** [review]
